March 13: Hashfunctions3 (#23)
cornelltech/CS6831-Spring2019: notes/hashfunctions3.tex

\section{PRFs from Hash functions}

\subsection{Other Hash based PRFs}
The Merkle Damg{\aa}rd construction from the previous section is a type of Prefix MAC (since the key is prepended to the message). Although the construction is secure PRF in the random oracle model, we saw how it is susceptible to length extension attacks. Several alternatives have been proposed to counter these attacks. 

> 

Reviewer: Define what hash^f is in the following constructions? Namely, the Merkle-Damgard domain extending transform with a compression function f.

\paragraph{Suffix MAC:}

> $\hash^f(M \Vert K)$ - The key is appended to the end instead of the beginning. Now, for a length extension attack to succeed, the adversary must find a collision for hash output until the key is added. To elaborate, if the adversary finds messages $M_0$ and $M_1$ such that $\hash^f(M_0) = \hash^f(M_1)$ by an offline collision attack, then it can break security for the overall construction without ever needing the key. To prove \PRF-security, therefore, we require that $\hash$ be collision resistant.

Reviewer: ```suggestion
$\hash^f(M \Vert K)$ - The key is appended to the end instead of the beginning. Now, for a length extension attack to succeed, the adversary must find a collision for the hash output of the message prefix up until the key is added. To elaborate, if the adversary finds messages $M_0$ and $M_1$ such that $\hash^f(M_0) = \hash^f(M_1)$ by an offline collision attack, then it can break security for the overall construction without ever needing the key. To prove \PRF-security, therefore, we require that $\hash$ be collision resistant.
```

\paragraph{Envelope MAC:} $\hash^f(K_1 \Vert M \Vert K_2)$ - A key $K_1$ is prepended and another key $K_2$ is appended to the input. The \PRF-security proof is similar to the one before except now collision resistance is no longer required.

\paragraph{Nested MAC:}
$\hash^f(K_1 \Vert \hash^f(K_2 \Vert M))$ - The standard Prefix MAC construction is applied to the message twice with two \textit{independently} chosen keys $K_1$ and $K_2$.

\subsection{The HMAC Construction}

> $\HMAC$, a widely used MAC in practice is a close variant of the nested MAC. For $\HMAC$, The keys $K_1$ and $K_2$ are no longer picked independently but rather derived from some base key $K$. The keys are computed as $K_1 = f(\IV, K \oplus \ipad)$ and $K_2 = f(\IV, K \oplus \opad)$ where $\ipad$ (inner pad) and $\opad$ (outer pad) are constants. Figure \ref{fig: HMAC construction} shows both a shortened and an extended version of the $\HMAC$ construction

Reviewer: ```suggestion
$\HMAC$, a widely used MAC in practice, is a close variant of the nested MAC. For $\HMAC$, The keys $K_1$ and $K_2$ are no longer picked independently but rather derived from some base key $K$. The keys are computed as $K_1 = f(\IV, K \oplus \ipad)$ and $K_2 = f(\IV, K \oplus \opad)$ where $\ipad$ (inner pad) and $\opad$ (outer pad) are constants. Figure \ref{fig: HMAC construction} shows both a shortened and an extended version of the $\HMAC$ construction
```

Reviewer: Explain why ipad and opad are used. My understanding is so that HMAC can be built with SHA-256 black boxes without modifying IV.

    \modify{If $\phi \in \Phi$ then}\\
    \modify{\myInd Ret $\bot$}\\
    Ret $f(X,\modify{\phi(K)})$\medskip
}
{
    \underline{$\LRKAPRFReal^\advA_{f, \modify{\Phi}}$}\\[1pt]
    $K \getsr \bits^d$\\
    $\rho \getsr \Func(\modify{\bits^n \times \bits^d},n)$\\
    $b' \getsr \advA^{\FnOracle}$\\
    Ret $b'$\medskip

    \underline{$\FnOracle(\modify{\phi}, X)$}\\
    \modify{If $\phi \in \Phi$ then}\\
    \modify{\myInd Ret $\bot$}\\
    Ret $\rho(X \modify{, \phi(K)})$\medskip
}
\caption{Related Key Attack Game} \label{fig:RKA}
\end{figure}

We note that this game modification is far more general than what might be needed for HMAC security. For our application, it is reasonable to restrict $\Phi$ to just $\phi_{\ipad}$ and $\phi_{\opad}$ where $\phi_{\ipad}(K) = K \oplus \ipad$ and $\phi_{\opad}(K) = K \oplus \opad$ and then ask whether $f$ is \RKAPRF secure. Choosing such an $f$ function would imply that $K_1$ and $K_2$ are indistinguishable from random bit strings. 

Suppose now that we have ensured that the compression function $f$ used in HMAC is \RKAPRF secure. We still need to show that the overall construction is \PRF-secure. 

\begin{lemma}

> (Informal) If $f$ is $\RKAPRF$-secure and the iterated keyed hash function $\hash$ is collision resistant then the $\HMAC$ construction is $\PRF$-secure

Reviewer: Give formal lemma with advantage bounds, but then informal proof sketch?

\end{lemma}
\begin{proofsketch}
If $f$ is $\RKAPRF$-secure, then the keys $K_1$ and $K_2$ are indistinguishable from random bit strings even for adversaries that use outputs from related key functions $\phi_{\ipad}$ and $\phi_{\opad}$. Now, we have seen that the composition of a collision resistant function and a secure PRF is a secure PRF. Therefore, we can conclude that $\HMAC$ is PRF-secure
\end{proofsketch}

\noindent Some subtleties like block padding have been ignored in the above sketch. A detailed proof can be found in \cite{Bellare1996}. A stronger result that does not rely on the collision resistance of $\hash$ (but instead on some weaker assumptions) is proved in \cite{Bellare2006}

\subsection{The Indifferentiability Framework}

> Ideally, we want our hash functions to behave like random oracles, yet security analysis in the RO model is insufficient justification for instantiating constructions with real hash functions. Understanding security of real world constructions would be a lot easier if we had a general framework that accounted for structure-abusing attacks on hash functions. Coron et al. \cite{Coron2005} suggest use of the Indifferentiability franework introduced by Maurer et al. \cite{Maurer2004} in the context of hash functions. At a high level, we model the inner compression function $f$ itself as a random oracle for fixed-length inputs and attempt to prove that the overall construction is indistinguishable from a random oracle. Indifferentiability from a random oracle can be considered approximately as secure as a random oracle. 

Reviewer: ```suggestion
We've seen that security analysis in the RO model is insufficient justification for instantiating constructions with real hash functions, e.g., due to structure abusing attacks. Understanding security of real world constructions would be a lot easier if we had a general framework that accounted for structure-abusing attacks on hash functions. Ideally, we want our hash functions to behave like random oracles. Coron et al. \cite{Coron2005} suggest use of the Indifferentiability framework introduced by Maurer et al. \cite{Maurer2004} in the context of hash functions. At a high level, we model the inner compression function $f$ itself as a random oracle for fixed-length inputs and attempt to prove that the overall construction is indistinguishable from a random oracle. Indifferentiability from a random oracle can be considered approximately as secure as a random oracle. 
```

    - \Pr[\INDIFF0^\advD_{\Horacle,\mathcal{S}} \Rightarrow 1]
\right|
\]

\begin{wrapfigure}{r}{1.5in}
    \centering
    \fpage{.2}{
            \underline{$\textbf{adversary } \advD^{\FnOracle, O}$}\\[1pt]
            $m_1, m_2 \getsr \bits^d$\\
            $y_1 \gets \FnOracle(m_1)$\\
            $y_2 \gets O(y_1,  m_2)$\\
            $y'_2 \gets \FnOracle(m_1 \Vert m_2)$\\
            Ret $(y_2 = y'_2)$
    }
    \caption{MD adversary} \label{fig:Indiff distinguisher}
\end{wrapfigure}

\noindent First, we note that a good simulator $\simoracle$ for $f$ in the ideal world may not always exist. In fact, the Merkle Damg{\aa}rd construction is actually not indifferentiable from a random oracle. Consider the $\INDIFF$-adversary $\advD$ defined in Figure~\ref{fig:Indiff distinguisher}.
In the real world, the two outputs $y_2$ and $y'_2$ will always be equal. On the other hand, in the ideal world, the probability that $y_2$ and $y'_2$ are equal is just the probability that a randomly sampled $n$ bit string is equal to the $y_2$ output by the simulator; this is equal to $\frac{1}{2^n}$. Therefore, $\advD$ distinguishes between the real and ideal worlds with high probability.

> Variations of the Merkle Damg{\aa}rd construction such as the Chopped MD or the Enveloped MD are often used in practice instead since they are indifferentiable from a random oracle

Reviewer: Maybe also mention sponge construction?

\bigskip
\noindent How is this indifferentiable framework useful? It's real power is seen through the following composition theorem which allows security analysis of a construction that uses the hash function $\hash^f$ for any arbitrary game $G$

\begin{theorem}
\label{thm: indiff composition}
\textbf{Indifferentiability Composition Theorem} \cite{Maurer2004}
For any security game $G$, construction $\hash^f$ and scheme $C$, if 
\begin{enumerate}
    \item $\hash^f$ is indifferentiable from a random oracle under the assumption that $f$ is a random oracle
    \item C is provably $G$-secure in the random oracle model
\end{enumerate}

> then, $C$ is provably $G$-secure using $\hash^f$ under the assumption that $f$ is a random oracle

Reviewer: I know this is how Tom stated the thm in the lecture slides, but might be more formal to specify what G-secure and indifferentiable mean by stating the theorem in terms of an advantage bound. Something like:
Pr[G^{\advA, RO} =>1] \leq Pr[G^{\advA, \hash}  => 1] + \AdvIndiff^{\advB}

Reviewer: Can check with Tom to see if that makes sense...
